File: busaninfoapp/RestaurantActivity.java
package com.example.busaninfoapp;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.annotation.SuppressLint;

import com.bumptech.glide.Glide;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.GoogleMap.OnMyLocationButtonClickListener;
import com.google.android.gms.maps.GoogleMap.OnMyLocationClickListener;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.Marker;
import com.google.android.gms.maps.model.MarkerOptions;

import android.Manifest;
import android.content.pm.PackageManager;
import android.location.Location;
import android.location.LocationManager;
import androidx.core.app.ActivityCompat;

import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;
import androidx.annotation.NonNull;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;

public class RestaurantActivity extends AppCompatActivity implements OnMapReadyCallback,
        OnMyLocationButtonClickListener,
        OnMyLocationClickListener {

    private static final int LOCATION_PERMISSION_REQUEST_CODE = 1;
    private Location lastLocation;
    private LocationManager locationManager;
    public View cardView;
    public TextView rName, rAddr, rMenu, rTel, rBusinessHours, rExplanation;
    // 사용할 이미지 뷰
    private ImageView food;
    private static final String jsonFile = "jsons/busanRestaurant.json";


    private static String[] PERMISSIONS =
            {Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION};

    final LatLng CITY_HALL = new LatLng(35.17914523506671, 129.07492106277513);
    private GoogleMap gMap;
    ArrayList<MyRestaurant> restaurants = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_restaurant);

        getSupportActionBar().setTitle("맛집 목록");

        SupportMapFragment mapFragment = (SupportMapFragment) getSupportFragmentManager()
                .findFragmentById(R.id.mapG);
        mapFragment.getMapAsync(this);

    }

    @Override
    public void onMapReady(final GoogleMap googleMap) {
        gMap = googleMap;
        if (getLocationPermission()) {
            initMap(gMap);
        }
        else {
            ActivityCompat.requestPermissions(this, PERMISSIONS, LOCATION_PERMISSION_REQUEST_CODE);
        }
        cardView = findViewById(R.id.cardRestaurant);

        rName = findViewById(R.id.RestNametxt);
        rBusinessHours = findViewById(R.id.businessHourstxt);
        rAddr = findViewById(R.id.RestAddrtxt);
        rMenu = findViewById(R.id.RestMenutxt);
        rTel = findViewById(R.id.RestTeltxt);
        rExplanation = findViewById(R.id.RestExptxt);
        // 이미지뷰랑 연결한 부분
        food = findViewById(R.id.foodImage);

        gMap.setOnMyLocationButtonClickListener(this);
        gMap.setOnMyLocationClickListener(this);

        gMap.setOnMapClickListener(new GoogleMap.OnMapClickListener() {
            @Override
            public void onMapClick(@NonNull LatLng latLng) {
                cardView.setVisibility(View.INVISIBLE);
            }
        });

        gMap.setOnInfoWindowClickListener(new GoogleMap.OnInfoWindowClickListener() {
            @Override
            public void onInfoWindowClick(@NonNull Marker marker) {
                MyRestaurant item = null;
                for (int i = 0; i < restaurants.size(); i++) {
                    if(restaurants.get(i).findIndex(marker.getTitle()) != -1) {
                        item = restaurants.get(i);
                    }
                }
                // 이미지 주소 받은 부분
                String imageUrl = item.getImageurl();

                rName.setText(item.getTitle());
                String title = item.getTitle();
                rBusinessHours.setText("영업시간:  " + item.getBusinessHours());
                rAddr.setText("주소:  " + item.getAddr());
                rMenu.setText("메뉴:  " + item.getMenu());
                rTel.setText("전화번호:  " + item.getTelNum());
                rExplanation.setText("가게 설명:  " + item.getExplanation());
                Glide.with(RestaurantActivity.this).load(imageUrl).into(food);
                cardView.setVisibility(View.VISIBLE);
                Intent it = new Intent(RestaurantActivity.this, RestaurantReviewActivity.class);
                findViewById(R.id.reviewButton).setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        it.putExtra("name", title);
                        startActivity(it);
                    }
                });
            }
        });

    }

    // 퍼미션 확인부분
    private Boolean getLocationPermission() {
        for (int i = 0; i < 2; i++) {
            String permission = PERMISSIONS[i];
            if (ActivityCompat.checkSelfPermission(this.getApplicationContext(), permission) != PackageManager.PERMISSION_GRANTED) {
                return false;
            }
        }
        return true;
    }

    @SuppressLint("MissingPermission")
    public void initMap(GoogleMap googleMap) {
        if (getLocationPermission()) {
            googleMap.setMyLocationEnabled(true);
            googleMap.getUiSettings().setMyLocationButtonEnabled(true);
            addMarker(googleMap);
            googleMap.moveCamera(CameraUpdateFactory.newLatLngZoom(getMyLocation(), 15));
        }
        else {
            googleMap.moveCamera(CameraUpdateFactory.newLatLngZoom(CITY_HALL, 15));
        }


    }

    @SuppressLint("MissingPermission")
    public LatLng getMyLocation() {
        String locationProvider = LocationManager.GPS_PROVIDER;
        locationManager = (LocationManager) this.getSystemService(Context.LOCATION_SERVICE);
        lastLocation = locationManager.getLastKnownLocation(locationProvider);
        if (lastLocation != null) {
            return new LatLng(lastLocation.getLatitude(), lastLocation.getLongitude());
        }
        else {
            return CITY_HALL;
        }
    }



    @Override
    public void onRequestPermissionsResult(int requestCode,
                                           @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }
    @Override
    public boolean onMyLocationButtonClick() {
        Toast.makeText(this, "MyLocation button clicked", Toast.LENGTH_SHORT).show();
        // Return false so that we don't consume the event and the default behavior still occurs
        // (the camera animates to the user's current position).
        return false;
    }

    @Override
    public void onMyLocationClick(@NonNull Location location) {
        Toast.makeText(this, "Current location:\n" + location, Toast.LENGTH_LONG).show();
    }


    public JSONArray loadJson(String jsonFile) {
        String bf = "";

        JSONArray restaurant = null;
        try {
            InputStream is = getAssets().open(jsonFile);

            int FileSize = is.available();
            byte[] buffer = new byte[FileSize];
            is.read(buffer);
            is.close();

            bf = new String(buffer, "UTF-8");

            JSONObject jsonObject = new JSONObject(bf);

            restaurant = jsonObject.getJSONArray("item");

        } catch (IOException ex) {
            ex.printStackTrace();
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return restaurant;
    }

    public void addMarker(GoogleMap googleMap) {
        JSONArray restaurant = loadJson(jsonFile);
        try {
            for (int i = 0; i < restaurant.length(); i++) {
                JSONObject rest = restaurant.getJSONObject(i);
                MyRestaurant item = new MyRestaurant(rest.getInt("UC_SEQ"), rest.getString("TITLE"), rest.getString("ADDR1"), rest.getString("CNTCT_TEL")
                        , rest.getString("RPRSNTV_MENU"), rest.getString("USAGE_DAY_WEEK_AND_TIME"), rest.getString("ITEMCNTNTS")
                        , rest.getDouble("LAT"), rest.getDouble("LNG"), rest.getString("MAIN_IMG_NORMAL"));
                MarkerOptions markerOptions = new MarkerOptions();
                markerOptions.position(new LatLng(item.getLat(), item.getLng()));
                markerOptions.title(item.getTitle());
                markerOptions.snippet(item.getMenu());

                googleMap.addMarker(markerOptions);
                restaurants.add(item);
            }
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }
}

File: test/RestaurantActivity.kt
package com.example.test

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.Manifest
import android.annotation.SuppressLint
import android.content.pm.PackageManager
import android.location.Location
import android.location.LocationManager
import android.widget.Toast
import androidx.core.app.ActivityCompat;
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.BitmapDescriptorFactory
import com.google.maps.android.clustering.ClusterManager
import android.content.Context
import android.graphics.Bitmap
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.text.TextUtils
import android.widget.ArrayAdapter
import androidx.core.content.res.ResourcesCompat
import com.google.android.gms.maps.model.MarkerOptions
import kotlinx.android.synthetic.main.activity_restaurant.*;
import org.json.JSONArray
import org.json.JSONObject
import java.net.URL

class RestaurantActivity : AppCompatActivity() {
    // 권한 필요한 퍼미션 목록
    val PERMISSIONS = arrayOf(
        Manifest.permission.ACCESS_COARSE_LOCATION,
        Manifest.permission.ACCESS_FINE_LOCATION
    )

    // 퍼미션 승인 요청시 사용하는 코드
    val REQUEST_PERMISSION_CORE = 1

    // 기본 맵 줌 레벨
    val DEFAULT_ZOOM_LEVEL = 14f

    // 위치 정보를 불러올 수 없는 경우 부산 시청의 위치 정보를 가져와 화면에 표시
    val CITY_HALL = LatLng(35.17914523506671, 129.07492106277513)

    // 구글 맵 불러오기
    val googleMap: GoogleMap? = null
    val itemMap = mutableMapOf<JSONObject, MyItem>()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_restaurant)

        mapView.onCreate(savedInstanceState)



        if(hasPermissions()) {
            initMap()
        }
        else {
            ActivityCompat.requestPermissions(this, PERMISSIONS, REQUEST_PERMISSION_CORE)
        }

    }

    // 사용 권한에 확인 메서드
    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>,
        grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)

        initMap()

    }

    fun hasPermissions(): Boolean {

        for(permission in PERMISSIONS) {
            if(ActivityCompat.checkSelfPermission(this, permission) != PackageManager.PERMISSION_GRANTED) {
                return false
            }
        }
        return true
    }


    // 구글 맵 불러오기
    @SuppressLint("MissingPermission")
    fun initMap() {

        mapView.getMapAsync {


            it.uiSettings.isMyLocationButtonEnabled = false
            when {
                hasPermissions() -> {
                    it.isMyLocationEnabled = true
                    it.moveCamera(CameraUpdateFactory.newLatLngZoom(getMyLocation(), DEFAULT_ZOOM_LEVEL))
                    // 지도에 마커 추가하는 부분
                    val json = assets.open("jsons/busanFood.json").reader().readText()
                    val jsonArray = JSONArray(json)
                    for (i in 0 until jsonArray.length()) {
                        val restaurant = jsonArray.getJSONObject(i)
                        val lat = restaurant.getDouble("LAT")
                        val lng = restaurant.getDouble("LNG")
                        val mark = LatLng(lat, lng)
                        it.addMarker(
                            MarkerOptions().position(mark)
                                .title(restaurant.getString("MAIN_TITLE"))
                                .snippet(restaurant.getString("RPRSNTV_MENU"))
                        )
                    }
                }
                else -> {
                    it.moveCamera(CameraUpdateFactory.newLatLngZoom(CITY_HALL, DEFAULT_ZOOM_LEVEL))
                }
            }
        }
    }

    // 현재 위치 가져오기
    @SuppressLint("MissingPermission")
    fun getMyLocation(): LatLng{
        val locationProvider: String = LocationManager.GPS_PROVIDER
        val locationManager = getSystemService(Context.LOCATION_SERVICE) as LocationManager
        val lastLocation: Location? = locationManager.getLastKnownLocation(locationProvider)
        if(lastLocation != null) {
            return LatLng(lastLocation.latitude, lastLocation.longitude)
        }
        else {
            return CITY_HALL
        }
    }

    fun searchMyLocation() {
        when {
            hasPermissions() -> googleMap?.moveCamera(CameraUpdateFactory.
            newLatLngZoom(getMyLocation(), DEFAULT_ZOOM_LEVEL))
            else -> Toast.makeText(applicationContext, "위치사용권한 설정에 동의해주세요", Toast.LENGTH_LONG).show()
        }
    }

    override fun onResume() {
        super.onResume()
        mapView.onResume()
    }

    override fun onPause() {
        super.onPause()
        mapView.onPause()
    }

    override fun onDestroy() {
        super.onDestroy()
        mapView.onDestroy()
    }

    override fun onLowMemory() {
        super.onLowMemory()
        mapView.onLowMemory()
    }


    override fun onStart() {
        super.onStart()
        LocationSerButton.setOnClickListener { searchMyLocation()}
    }
}










































